doc/source/programs/gdalwarp.rst

```diff
@@ -493,6 +502,164 @@ The details of how it is taken into account depends on the resampling kernel:
+threshold then the remaining points are approximated (in two chunks utilizing
+the center point). If the error exceeds the threshold, the scanline is split
+into two sections, and the approximator is recursively applied to each section
+until the error is less than the threshold or all points have been exactly
+computed.
+
+The error threshold (in pixels) can be controlled with the gdalwarp
+:option:`-et` switch. If you want to compare a true pixel-by-pixel reprojection
+use :option:`-et 0` which disables this approximator entirely.
+
+
+Memory usage
+------------
+
+Adding RAM will almost certainly increase the speed of :program:`gdalwarp`.
+That's not at all the same as saying that it is worth it, or that the speed
+increase will be significant. Disks are the slowest part of the process.
+By default :program:`gdalwarp` won't take much advantage of RAM. Using the flag
+:option:`-wm 500` will operate on 500MB chunks at a time which is better than the
+default. Increasing the I/O block cache size may also help. This can be done by
+setting the :config:`GDAL_CACHEMAX` configuration like:
+
+.. code-block:: bash
+
+   gdalwarp --config GDAL_CACHEMAX 500 -wm 500 ...
+
+This uses 500MB of RAM for read/write caching, and 500MB of RAM for working
+buffers during the warp. Beyond that it is doubtful more memory will make a
+substantial difference.
+
+Check CPU usage while :program:`gdalwarp` is running. If it is substantially
+less than 100% then you know things are IO bound. Otherwise they are CPU bound.
+The ``--debug`` option may also provide useful information. For instance, after
+running the following:
+
+.. code-block::
+
+   gdalwarp --debug on abc.tif def.tif
+
+a message like the following will be output:
+
+::
+
+  GDAL: 224 block reads on 32 block band 1 of utm.tif
+
+In this case it is saying that band 1 of :file:`utm.tif` has 32 blocks, but
+that 224 block reads were done, implying that lots of data was having to be
+re-read, presumably because of a limited IO cache. You will also see messages
+like:
+
+::
+
+   GDAL: GDALWarpKernel()::GWKNearestNoMasksByte()
+   Src=0,0,512x512 Dst=0,0,512x512
+
+The Src/Dst windows show you the "chunk size" being used. In this case my whole
+image which is very small. If you find things are being broken into a lot of
+chunks increasing :option:`-wm` may help somewhat.
+
+But far more important than memory are ensuring you are going through an
+optimized path in the warper. If you ever see it reporting
+``GDALWarpKernel()::GWKGeneralCase()`` you know things will be relatively slow.
+Basically, the fastest situations are nearest neighbour resampling on 8bit data
+without nodata or alpha masking in effect.
+
+.. caution::
+
+   Increasing the value of the :option:`-wm` param may lead to loss of
+   performance in certain circumstances, particularly when gdalwarping *many*
+   small rasters into a big mosaic. See http://trac.osgeo.org/gdal/ticket/3120
+   for more details
+
```

Comment:
My read of the linked ticket was that the issue is resolved and this note can be removed.

Reply:
yes.

those notes also are quite ancient, and since then, multi-threading warping is available. Increasing the warp memory makes more sense if you also add --config GDAL_NUM_THREADS ALL_CPUS
A bigger value of -wm alone hardly helps. People are always keen to increase such thumbs since they believe "the bigger the better", but I'm quite skeptical it significantly helps in most cases.

Reply:
> Increasing the warp memory makes more sense if you also add --config GDAL_NUM_THREADS ALL_CPUS

So would it be correct to say "the warp memory specified by -wm is shared among all threads, so it is especially beneficial to increase this value when running gdalwarp with GDAL_NUM_THREADS > 1" ?

Also, over on the config option page `GDAL_NUM_THREADS` has a default of `ALL_CPUS` but I see in `GWKThreadsCreate` it defaults to 1. So I guess that doc should say the default depends on the context?

Reply:
> So would it be correct to say "the warp memory specified by -wm is shared among all threads, so it is especially beneficial to increase this value when running gdalwarp with GDAL_NUM_THREADS > 1" ?

sort of... actually what happens is that the number of lines of each warping chunk is split among the threads available. If the warping chunk is too small, the setup cost of the threads can be non neglectable.
Let's settle with your proposed phrasing, as I can't find something better.

> Also, over on the config option page GDAL_NUM_THREADS has a default of ALL_CPUS but I see in GWKThreadsCreate it defaults to 1. So I guess that doc should say the default depends on the context?

yes

Reply:
Thanks. Updated with 

```
The warp memory specified by :option:`-wm` is shared among all
threads, so it is especially beneficial to increase this value when running
:program:`gdalwarp` with :option:`-wo NUM_THREADS` (or its equivalent
:config:`GDAL_NUM_THREADS`) greater than 1."
```